### PR TITLE
[SHELL] Folder Options: File Types: Implement CORE-12906

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
This PR improves Folder Options dialog.
JIRA issue: [CORE-12906](https://jira.reactos.org/browse/CORE-12906)

- Improved Folder Options codes.
- Added LoadStringDx function.
- Added Dialog IDD_NEW_EXT "Create New Extension" and Dialog IDD_EDIT_FILE_TYPE "Edit File Type".
- Added "New", "Delete" and "Advanced" actions.

Some icon images are wrong or invalid.

### CODE REVIEW

**Comment by katahiromz** [comment]
What should we do if there's no icon for one item?

**Comment by HBelusca** [comment]
I think we should display the default blank icon (it looks like an empty file).

**Comment by katahiromz** [comment]
Fixed icon images.

**Comment by learn-more** [comment]
Is this PR a work in progress @katahiromz ?

**Comment by katahiromz** [comment]
It's done.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 1]
```diff
@@ -1388,12 +1392,71 @@ InsertFileType(HWND hDlgCtrl, WCHAR * szName, PINT iItem, WCHAR * szFile)
...
     /* close key */
     RegCloseKey(hKey);
 
+    INT iLargeImage = -1, iSmallImage = -1;
+    if (Entry->hIconLarge && Entry->hIconSmall)
+    {
+        iLargeImage = ImageList_AddIcon(himlLarge, Entry->hIconLarge);
+        iSmallImage = ImageList_AddIcon(himlSmall, Entry->hIconSmall);
+    }
+    assert(iLargeImage == iSmallImage);
```
Comment (binarymaster) [inline dll/win32/shell32/dialogs/folder_options.cpp]: Shouldn't an `ASSERT` be upper-case?

**Comment by katahiromz** [comment]
Do we need the large image list?

**Comment by JaredSmudde** [comment]
untitled
@katahiromz You need it. If you click Advance, it has a 32 pixel icon.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: I have no idea that which is the better.

Comment (ThFabba) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Yeah, ASSERT is preferred. assert() comes from the CRT and will give you a message box, while ASSERT goes to the kernel debugger, so it's better for testbot etc.

Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Done.

Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Done.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 2]
```diff
@@ -1388,12 +1392,71 @@ InsertFileType(HWND hDlgCtrl, WCHAR * szName, PINT iItem, WCHAR * szFile)
...
     /* close key */
     RegCloseKey(hKey);
 
+    INT iLargeImage = -1, iSmallImage = -1;
+    if (Entry->hIconLarge && Entry->hIconSmall)
+    {
+        iLargeImage = ImageList_AddIcon(himlLarge, Entry->hIconLarge);
+        iSmallImage = ImageList_AddIcon(himlSmall, Entry->hIconSmall);
+    }
+    ASSERT(iLargeImage == iSmallImage);
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: I would perhaps move this ASSERT inside the `if` block.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 3]
```diff
@@ -1525,6 +1610,130 @@ FindSelectedItem(
...
+    const UINT Flags = LVNI_ALL | LVNI_SELECTED;
+    INT iItem = ListView_GetNextItem(hListView, -1, Flags);
+    if (iItem == -1)
+        return;
+
+    PFOLDER_FILE_TYPE_ENTRY pItem = FindSelectedItem(hListView, iItem);
+    if (pItem == NULL)
+        return;
+
+    // FIXME: registry
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: What needs to be done in addition?

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Done.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: HKEY_CLASSES_ROOT contains many file type informations. For example, there is ".txt" registry key. And "txtfile" registry key also exists. If the user had requested deletion, we have to delete both of ".txt" and "txtfile".

Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Implemented file type deletion.

**Comment by katahiromz** [comment]
Get it go.

**Comment by katahiromz** [comment]
I had deleted my repository. I can't edit.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 4]
```diff
@@ -1525,6 +1610,172 @@ FindSelectedItem(
     return NULL;
 }
 
+INT_PTR CALLBACK
+NewExtDlgProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    /* FIXME */
+    static HWND s_hListView = NULL;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: It's a dialog procedure to add new file extension info.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 5]
```diff
@@ -1525,6 +1610,172 @@ FindSelectedItem(
...
+    HeapFree(GetProcessHeap(), 0, pItem);
+
+    InvalidateRect(hListView, NULL, TRUE);
+}
+
+INT_PTR CALLBACK
+EditFileTypeDlgProc(HWND hwndDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    /* FIXME */
+    static PFOLDER_FILE_TYPE_ENTRY s_pItem = NULL;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: For future use.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 6]
```diff
@@ -1525,6 +1610,172 @@ FindSelectedItem(
...
+    static HICON s_hIcon = NULL;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: We can delete it.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 7]
```diff
@@ -1525,6 +1610,172 @@ FindSelectedItem(
...
+void FileTypesDlg_OnDelete(HWND hwndDlg, HWND hListView)
+{
+    const UINT Flags = LVNI_ALL | LVNI_SELECTED;
+    INT iItem = ListView_GetNextItem(hListView, -1, Flags);
+    if (iItem == -1)
+        return;
+
+    PFOLDER_FILE_TYPE_ENTRY pItem = FindSelectedItem(hListView, iItem);
+    if (pItem == NULL && pItem->FileExtension[0] == UNICODE_NULL)
+        return;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Quick deletion of file extension info without whole view update is a required feature.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: We need accessibility to list view in the procedure.

[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: For future use.

**Comment by HBelusca** [comment]
There must be a way to retrieve the code of a git PR so that you can reattach it to a branch.

**Comment by ThFabba** [comment]
git fetch https://github.com/reactos/reactos refs/pull/316/head:some-branch
git checkout some-branch

However I don't know whether/how you can re-attach a new branch to this PR.

**Comment by sanchaez** [comment]
I guess it's best to make a diff and then re-apply it onto a new branch.
Closing PR as non-mergeable, reopen if I'm wrong here.

**Comment by HBelusca** [comment]
Superseded by PR #544.

**Pull request closed without merging**